README.md
# Highkara news reader

Highkara news reader helps you to follow news from Finland and other countries. It's an unofficial client for High.fi news aggregator and provides functionality for reading news from several medias and regions. The app is similar to Ampparit

Highkara is available both for iPhone and iPad with iOS 8 or newer operating system. The application is open source and licensed under MIT.

## Technologies

* Swift 2
* Alamofire

# Building

XCode to build the app.

[CocoaPods](https://github.com/CocoaPods) is used to manage dependencies.

## Dependencies

[Alamofire](https://github.com/Alamofire/Alamofire) is used for JSON requests and responses.



changelog.txt
0.13.0 (2019-09-09)
* Updated Alamofire dependencies
* Add missing new icons

0.12.0 (2018-03-31)
* Add option to show paid-for / paywalled news sites which are omitted by default
* Migrate to Swift 4
* Fix minor UI layout issues in settings
* Remove exception domains as High.fi provides now https

0.11.3 (2017-12-23)
* Add support for iPhone X full screen mode by adding correct launch image.
* Fix problem in High.fi API when articleID is String and not Int (at least in media, tiede, netflix-uudet ja viranomaiset) by parsing articleID from Any to Int.
* Remove GoogleToolboxForMac from Pods
* Use case success and failure for Alamofire json

0.11.2 (2017-11-02)
* Guard againts decoding values from Swift 2 and Swift 3
* Use GoogleAnalytics as Google/Analytics is deprated

0.11.1 (2017-11-01)
* Add support for AMP URLs
* Add support for extended Today Widget
* Fix some layout issues with Today Widget
* Update Alamofire libraries

0.11.0 (2017-06-04)
* Add WatchOS extension app
* Show category menu on iPad with swipe

0.10.1 (2017-04-29)
* Remove old version checks for <iOS 9
* Increase category indentation for clarity
* Fix minor bugs

0.10.0 (2017-04-20)
* Add new user specific category (ICYMI, In case you missed)
* Update API calls to match new requirements
* Add accessibility
* Fix bugs: no news picture, search paging

0.9.14 (2016-11-11)
* Add Search functionality
* Fix layout issues with iPad in Settings view

0.9.13 (2016-11-03)
* Fix layout issues in Settings view
* Move browser related settings to own section
* Add option to open links in new tab if Chrome used

0.9.12 (2016-10-02)
* Refactor texts to use system text size
* Add option to use Chrome for opening news
* Add option to select used text size

0.9.11 (2016-09-26)
* Fix Today widget color in iOS 10
* Remove news title hilighting for now
* Update libraries and project to Swift 2.3 and Xcode 8

0.9.10 (2016-09-14)
* Added option to show news picture with news
* Added hilighting the news title
* Fixed news pagination
* Added Google Analytics for helping to develop the app
* Updated to Alamorefire 3 and AlamofireImage 2

0.9.9 (2016-08-26)
* Added support of Today Widget also in iOS 8
* Refactored About screen to use tableview
* Fix theme refreshed event in Settings

0.9.8 (2016-08-22)
* Added Today widget for easier following of news, possibility to change its category from Settings screen. 
* Reworked Settings screen to use tableview.
* Added Netherlands and Denmark to exception domains.

0.9.7
* Skipped

0.9.6 (2016-06-15)
* Add new category for showing combined favorites in one list
* Add publishedtime to news entry.
* Don't clear news list when getting next page. 
* Add alert dialog to ask if resetting settings.

0.9.5 (2016-02-16)
* Add search for filtering news sources
* Move region selection to textfield with picker as input
* Add 3D Touch to peek in category view and news view
* Add new actions for news item: filter, open in external browser

0.9.3 (2016-01-16)
* Add darker dark theme
* Adjust table views UI: no separators, clock icon
* Fix issue with landscape layout

0.9.2 (2015-12-09)
* Fix sharing news.
* Fix UI issues with settings table views.

0.9.1 (2015-12-07)
* Add possibility to hide categories from news back to settings.
* Adjust even row color in light theme.
* Adjust swipe to dismiss gesture in SFSafariViewController
* Don't show descriptions by default.

0.9.0 (2015-11-30)
* Show full title in news list.
* Dark theme.
* Add API methods to list news sources. 
* Switch hiding categories feature to filtering news sources.
* Filter shown news by SourceID when getting news from API.
* Fix top category grouping which isn't by time but just numbers.
* Add swipe from edge gesture to SFSafariViewController until it comes in iOS 9.2.

Haikara/DataModels.swift
//
//  DataModels.swift
//  Haikara
//
//  Created by Marko Wallin on 27.6.2015.
//  Copyright (c) 2015 Rule of tech. All rights reserved.
//
//
//  News.swift
//  NewsPages
//
//  Created by Marko Wallin on 21.6.2015.
//  Copyright (c) 2015 Rule of tech. All rights reserved.
//

import UIKit

// listLanguages
// http://high.fi/api/?act=listLanguages&APIKEY=123
// { "responseData": { "supportedLanguages": [ { "language": "Finnish", "country": "Finland", "domainToUse": "high.fi", "languageCode": "fi-fi", "mostPopularName": "Suosituimmat", "latestName": "Uusimmat", "useToRetrieveLists": "finnish", "genericNewsURLPart": "uutiset" }, { "language": "English", "country": "United States", "domainToUse": "en.high.fi", "languageCode": "en-us", "mostPopularName": "Most Popular", "latestName": "Latest News", "useToRetrieveLists": "english", "genericNewsURLPart": "news" } ] } }

class Language: NSObject, NSCoding {
    let language: String
    let country: String
    let domainToUse: String
    let languageCode: String
    let mostPopularName: String
    let latestName: String
    let useToRetrieveLists: String
    let genericNewsURLPart: String
    
    init(language: String, country: String, domainToUse: String, languageCode: String,
        mostPopularName: String, latestName: String, useToRetrieveLists: String, genericNewsURLPart: String) {
        self.language = language
        self.country = country
        self.domainToUse = domainToUse
        self.languageCode = languageCode
        self.mostPopularName = mostPopularName
        self.latestName = latestName
        self.useToRetrieveLists = useToRetrieveLists
        self.genericNewsURLPart = genericNewsURLPart
        super.init()
    }
    
    required init(coder aDecoder: NSCoder) {
        language = aDecoder.decodeObjectForKey("language") as! String
        country = aDecoder.decodeObjectForKey("country") as! String
        domainToUse = aDecoder.decodeObjectForKey("domainToUse") as! String
        languageCode = aDecoder.decodeObjectForKey("languageCode") as! String
        mostPopularName = aDecoder.decodeObjectForKey("mostPopularName") as! String
        latestName = aDecoder.decodeObjectForKey("latestName") as! String
        useToRetrieveLists = aDecoder.decodeObjectForKey("useToRetrieveLists") as! String
        genericNewsURLPart = aDecoder.decodeObjectForKey("genericNewsURLPart") as! String
    }
    
    func encodeWithCoder(aCoder: NSCoder) {
        aCoder.encodeObject(language, forKey: "language")
        aCoder.encodeObject(country, forKey: "country")
        aCoder.encodeObject(domainToUse, forKey: "domainToUse")
        aCoder.encodeObject(languageCode, forKey: "languageCode")
        aCoder.encodeObject(mostPopularName, forKey: "mostPopularName")
        aCoder.encodeObject(latestName, forKey: "latestName")
        aCoder.encodeObject(useToRetrieveLists, forKey: "useToRetrieveLists")
        aCoder.encodeObject(genericNewsURLPart, forKey: "genericNewsURLPart")
    }
    
    override var description: String {
        return "Language: language=\(self.language), country=\(self.country), domainToUse=\(self.domainToUse), languageCode=\(self.languageCode), mostPopularName=\(self.mostPopularName), latestName=\(self.latestName), useToRetrieveLists=\(self.useToRetrieveLists), genericNewsURLPart=\(self.genericNewsURLPart)"
    }
}

// listCategories
// http://fi.high.fi/api/?act=listCategories&usedLanguage=finnish
// { "responseData": { "categories": [ { "title": "Kotimaa", "sectionID": 95, "depth": 1, "htmlFilename": "kotimaa" }, { "title": "Ulkomaat", "sectionID": 96, "depth": 1, "htmlFilename": "ulkomaat" }, { "title": "Talous", "sectionID": 94, "depth": 1, "htmlFilename": "talous" }, { "title": "Urheilu", "sectionID": 98, "depth": 1, "htmlFilename": "urheilu" } ] } }

class Category: NSObject, NSCoding {
    let title: String
    let sectionID: Int
    let depth: Int
    let htmlFilename: String
    var selected: Bool
    
    required init(title: String, sectionID: Int, depth: Int, htmlFilename: String, selected: Bool) {
        self.title = title
        self.sectionID = sectionID
        self.depth = depth
        self.htmlFilename = htmlFilename
        self.selected = selected
        super.init()
    }
    
    required init(coder aDecoder: NSCoder) {
        title = aDecoder.decodeObjectForKey("title") as! String
        sectionID = aDecoder.decodeObjectForKey("sectionID") as! Int
        depth = aDecoder.decodeObjectForKey("depth") as! Int
        htmlFilename = aDecoder.decodeObjectForKey("htmlFilename") as! String
        selected = aDecoder.decodeObjectForKey("selected") as! Bool
    }
    
    func encodeWithCoder(aCoder: NSCoder) {
        aCoder.encodeObject(title, forKey: "title")
        aCoder.encodeObject(sectionID, forKey: "sectionID")
        aCoder.encodeObject(depth, forKey: "depth")
        aCoder.encodeObject(htmlFilename, forKey: "htmlFilename")
        aCoder.encodeObject(selected, forKey: "selected")
    }
    
    override var description: String {
        return "Category: title=\(self.title), sectionID=\(self.sectionID), depth=\(self.depth), htmlFilename=\(self.htmlFilename), selected=\(self.selected)"
    }
}

//{
//	"responseData": {
//		"feed": {
//			"title": "HIGH.FI",
//			"link": "http://fi.high.fi/",
//			"author": "AfterDawn Oy",
//			"description": "News",
//			"type": "json",
//			"entries": [
//{
//	"title": "string",
//	"link": "url",
//	"author": "string",
//	"publishedDateJS": "2015-06-16T22:46:08.000Z",
//	"publishedDate": "June, 16 2015 22:46:08",
//	"originalPicture": "url",
//	"picture": "url",
//	"shortDescription": "",
//	"originalURL": "url",
//	"mobileLink": "",
//	"originalMobileURL": "",
//	"articleID": int,
//	"sectionID": int,
//	"sourceID": int,
//	"highlight": true
//
//},

class Entry: NSObject {
    let title: String
    let link: String
    let clickTrackingLink: String
    let author: String
    let publishedDateJS: String
    //	let picture: String?
    //	let originalPicture: String?
    var shortDescription: String?
    let originalURL: String
    var mobileLink: String?
    let originalMobileUrl: String?
    let shareURL: String
    var mobileShareURL: String?
    let articleID: Int
    var sectionID: Int
    let sourceID: Int
    let highlight: Bool
    var timeSince: String
    
    init(title: String, link: String, clickTrackingLink: String, author: String, publishedDateJS: String,
        shortDescription: String?, originalURL: String, mobileLink: String?, originalMobileUrl: String?, shareURL: String, mobileShareURL: String?, articleID: Int, sectionID: Int, sourceID: Int, highlight: Bool, timeSince: String) {
            self.title = title
            self.link = link
            self.clickTrackingLink = clickTrackingLink
            self.author = author
            self.publishedDateJS = publishedDateJS
            //	let picture: String?
            //	let originalPicture: String?
            self.shortDescription = shortDescription
            self.originalURL = originalURL
            self.mobileLink = mobileLink
            self.originalMobileUrl = originalMobileUrl
            self.shareURL = shareURL
            self.mobileShareURL = mobileShareURL
            self.articleID = articleID
            self.sectionID = sectionID
            self.sourceID = sourceID
            self.highlight = highlight
            self.timeSince = timeSince
    }
    
    override var description: String {
        return "Entry: title=\(self.title), link=\(self.link), clickTrackingLink=\(self.clickTrackingLink), author=\(self.author), published=\(self.publishedDateJS), desc=\(self.shortDescription), originalURL=\(self.originalURL), mobileLink=\(self.mobileLink), originalMobileUrl=\(self.originalMobileUrl), shareURL=\(self.shareURL), mobileShareURL=\(self.mobileShareURL), articleID=\(self.articleID), sectionID=\(self.sectionID), sourceID=\(self.sourceID), highlight=\(self.highlight), timeSince=\(self.timeSince)"
    }
}
